Add torch._utils.render_call, improve printoptions

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #102623

- Add get_printoptions and printoptions context manager
- Improve edgeitems handling when it is zero
- Add render_call which can be used to conveniently print command
  line arguments of a function call, while suppressing actual
  tensor data

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
Add torch._utils.render_call, improve printoptions

- Add get_printoptions and printoptions context manager
- Add render_call which can be used to conveniently print command
  line arguments of a function call, while suppressing actual
  tensor data

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

albanD commented on torch/_utils.py:
```diff
@@ -592,6 +592,25 @@ def dec(fun):
 
 
+def render_call(fn, args, kwargs):
+    import torch._tensor_str
```
I'm sure this file is already imported by default so that shouldn't be needed. And I don't expect this function to be called during import right?

albanD commented on torch/_utils.py:
```diff
@@ -592,6 +592,25 @@ def dec(fun):
 
+def render_call(fn, args, kwargs):
+    import torch._tensor_str
+    from torch.overrides import resolve_name
```
Is this torch/__init__.py import order issue? Why not just use `torch.overrides.resolve_name` directly below?

ezyang replied:
Mostly (1) I didn't want to assume that overrides was imported automatically from torch (it is, in fact) and (2) since this is a utils module I wanted to be sensitive to import order issues (but I don't think there are any)

albanD commented on torch/_utils.py:
```diff
@@ -592,6 +592,25 @@ def dec(fun):
+    if str_fn is None:
+        str_fn = str(fn)
+
+    def render_arg(t):
```
Do you need this indirection for local debugging? Otherwise we cna use repr directly below

ezyang replied:
In the old version of the patch when I wasn't changing printoptions, I had to tree_map over args/kwargs and knock out the Tensor arguments. I left this stub around in case we changed our mind. Not a big deal either way.

albanD reviewed (approved):
This is amazing!
Just small nits on the code

ezyang committed:
Update on "Add torch._utils.render_call, improve printoptions"


- Add get_printoptions and printoptions context manager
- Add render_call which can be used to conveniently print command
  line arguments of a function call, while suppressing actual
  tensor data

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge

ezyang commented:
@pytorchbot merge
